compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -401,7 +401,8 @@ class ClassfileParser(
       classRoot.setFlag(sflags)
       moduleRoot.setFlag(Flags.JavaDefined | Flags.ModuleClassCreationFlags)
 
-      val privateWithin = getPrivateWithin(jflags)
+      val jflags1 = innerClasses.get(currentClassName.toString).fold(jflags: Int)(_.jflags)
+      val privateWithin = getPrivateWithin(jflags1)
```

Comment:
Scala 2 sets the `privateWithin` for inner classes when creating the symbol in `enterOwnInnerClasses`

https://github.com/scala/scala/blob/v2.13.15/src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala#L1055-L1058

Maybe Scala 3 should also do it there?